### DISCUSSION TITLE
MySQL Lost connection and multithreaded

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#8891] [Usage Questions] [answered]
We experienced these errors, using `pymysql` driver + MySQL 5.4 instance, within a FastAPI REST application run by unicorn webserver.
FastAPI is used in a synchronous fashion (using the ThreadPool instead of Event Loop for any incoming HTTP request)

First we reach a connection overflow error : 

`sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00`

And we we restart the server, we immediatly get this error on next query to the DB :

`pymysql.err.OperationalError: (2013, 'Lost connection to MySQL server during query')`

I understand that our ConnectionPool (which is the default QueuePool) is trying to use already open TCP socket which are still in use when we restart the server. 

Am I wrong ? 

Two questions : 
1) Is there any way of checking socket status (using `connect` method) through existing implementations of connection pool classes (or should I create my own for that) ? 

2) If no TCP sockets are available : is there an existing implementation to increase dynamically the number of the Queue overflow ?

**Comment by zzzeek** [comment, marked as answer]
> 1. Is there any way of checking socket status (using connect method) through existing implementations of connection pool classes (or should I create my own for that) ?

this error has to do with pymysql's server conversation being in an invalid state, the "socket" is fine except to the degree that PyMySQL was passed an error code from the server (which means the socket is fine):

https://github.com/PyMySQL/PyMySQL/blob/dd47caae95011e79b9e2ee12549d23f05a7f839d/pymysql/connections.py#L701-L713

per most documentation this issue means a packet itself was too large https://stackoverflow.com/questions/10563619/error-code-2013-lost-connection-to-mysql-server-during-query and there are some settings that can be made to adjust for that.    

> 2. If no TCP sockets are available : is there an existing implementation to increase dynamically the number of the Queue overflow ?

there's not, and there's not a need for that, start with a pool size and an overflow that's appropriately sized for how much concurrency your web application allows.  if your web application allows for 500 simultaenous HTTP connections on a single Python process, your pool needs to be sized to grow as large as 500 connections.

beyond that, if you are getting this error you want to look at connections that are hung for some reason, a request taking too long, things like that.


Also note that we seem to be getting a flurry of concurrency issues involving FastAPI very suddenly, see for example https://github.com/MagicStack/asyncpg/issues/978 .      These errors, particularly the "lost connection" error, are often the side effect of conditions that were ultimately caused by concurrency issues.  FastAPI seems to be doing some dark magic with concurrency here so I would not rule out FastAPI .

**Reply by jossefaz** [reply]
Is there any existing implementation of connection Pool like the SingletonThreadPool that would not cleanup and close connections but rather could have a "release" or "dispose" method which would be called by the end of the transaction or "on-demand" once we know that the current thread will never need DB connection again. So, only one thread per connection like the SingletonThreadPool do using `thread.local()`, but once the transaction has ended, we could call a "release_conn" method and the same connection would be used for next thread that needs it

I think maybe on a thread safe version of the QueuePool which will lock connections each time a connection checkout, and release it once it checkin to the pool : maybe using events ?

**Reply by jossefaz** [reply]
From the docs : 
> Call [Engine.dispose()](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Engine.dispose) directly before the child process is created. This will also cause the child process to start with a new connection pool, while ensuring the parent connections are not transferred to the child process:

I solved the `pymysql.err.OperationalError: (2013, 'Lost connection to MySQL server during query')` error by calling `engine.dispose()` at server restart...
It solved this error, but I really do not know why ?

**Reply by CaselIT** [reply]
Queue pool keeps connections open. If you restart the server all connections in the pool will be stale, so the next time you go to use them an error similar to the one you reported will happen. At that point sqlalchemy will discard the connection, and create a new one the next time that connection is checked out of the pool.

If you dispose of the pool you are closing all existing connections, so a new one will be created when checking out a connection.

You may also enable pool_pre_ping to make sqlalchemy verify the connection on checkout.

**Reply by jossefaz** [reply]
How does those connections will everbe reused by SQLAlchemy ? It is a brand new python process (I did not fork it). So it should be brand new TCP sockets (file descriptors) no ?

**Reply by CaselIT** [reply]
I'm not sure I follow. You mentioned that you are using a fastapi server. That is a long lived process, there is no brand new python process when serving a request unless you are creating a Process explicitly

**Reply by zzzeek** [reply]
> Is there any existing implementation of connection Pool like the SingletonThreadPool that would not cleanup and close connections but rather could have a "release" or "dispose" method which would be called by the end of the transaction or "on-demand" once we know that the current thread will never need DB connection again. So, only one thread per connection like the SingletonThreadPool do using `thread.local()`, but once the transaction has ended, we could call a "release_conn" method and the same connection would be used for next thread that needs it

I wouldnt use SingletonThreadPool because it's not production quality.

if you want "every thread has a single connection", I would manage that yourself, using detached connections which are separated from any pool limitations, like this:

```py
def make_new_thread():
    # connection from the regular pool
    connection = engine.connect()

    # detach it! now this connection has nothing to do with the pool.
    connection.detach()

    # pass the connection to the thread.  
    t = new Thread(target=my_thread_runner, args=(connection, ...))
    t.start()

def my_thread_runner(connection):
    try:
        with Session(connection) as sess:
            # do my work 
    finally:
        # closes the connection, i.e. the socket etc.
        connection.close()
```
